staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go

```diff
@@ -158,25 +175,58 @@ func (q *delayingType) ShutDown() {
 	})
 }
 
+// Add sends an item for immediate queuing, preempting a waiting item as it is queued.
+func (q *delayingType) Add(item interface{}) {
+	q.addToDelayedQueue(item, 0, DelayedAddOptions{})
+}
+
 // AddAfter adds the given item to the work queue after the given delay
 func (q *delayingType) AddAfter(item interface{}, duration time.Duration) {
+	q.addToDelayedQueue(item, duration, DelayedAddOptions{})
+}
+
+// AddWithOptions sends an item for immediate queuing, being dropped if the same item is already waiting.
+func (q *delayingType) AddWithOptions(item interface{}, opts DelayedAddOptions) {
+	q.addToDelayedQueue(item, 0, opts)
+}
+
+// AddAfterWithOptions adds the given item to the work queue after the given delay with options.
+func (q *delayingType) AddAfterWithOptions(item interface{}, duration time.Duration, opts DelayedAddOptions) {
+	q.addToDelayedQueue(item, duration, opts)
+}
+
+// addToDelayedQueue contains the implementation of the different add methods for the queue.
+func (q *delayingType) addToDelayedQueue(item interface{}, duration time.Duration, opts DelayedAddOptions) {
 	// don't add if we're already shutting down
 	if q.ShuttingDown() {
 		return
 	}
 
-	q.metrics.retry()
-
-	// immediately add things with no delay
-	if duration <= 0 {
-		q.Add(item)
-		return
+	// Add the item into the waiting queue with the appropriate ready time
+	waitEntry := &waitFor{data: item, readyAt: q.clock.Now().Add(duration), dropIfWaiting: opts.DropIfWaiting}
+
+	// For Add() or AddAfter() with a 0 delay we must make the return synchronous with
+	// the item being processed into either the active or waiting queue.
+	// The client can override this behaviour with the new AsyncAdd option when they
+	// know it is safe to fire and forget.
+	var cb chan struct{}
+	if !opts.AsyncAdd && duration == 0 {
+		cb = make(chan struct{})
+		waitEntry.callBackCh = cb
 	}
```

Comment:
should this defaulting behavior be done in the functions that don't take `opts DelayedAddOptions`? otherwise, `AsyncAdd:false` is ineffective for immediate adds